### ISSUE TITLE
STRef inlining is broken

### ISSUE THREAD

**Issue by natefaubion** [purescript/purescript#3885]
## Description

STRefs should inline into local mutable variables.

## To Reproduce

Use an STRef, and observe the generated code.

## Expected behavior

STRefs are never inlined

## Additional context

The order of arguments in the pattern-matche for read/write/modify are not correct.
https://github.com/purescript/purescript/blob/1b91bda780d8d80906adb16a9d63472718cec1c0/src/Language/PureScript/CoreImp/Optimizer/MagicDo.hs#L138-L139

**Comment by hdgarrood** [comment]
cc @milesfrain, this is probably what you were seeing in the book

**Comment by natefaubion** [comment]
My original report is slightly wrong. It is inlined sometimes, but I can't tell when.

**Comment by natefaubion** [comment]
I guess it does not work if you are using things like `while`, which is when you really need it most.

**Comment by natefaubion** [comment]
Oh! It's because I used `STArray.run`. If you use anything like that, and not ST.run, then inlining never happens for vars.

**Comment by natefaubion** [comment]
I think there are two ways to resolve this:
* Don't export `run`-like utilities for other ST types
* Account for all of these utilities in the inliner.

I'm personally leaning towards the first one, personally.

**Comment by hdgarrood** [comment]
The nice thing about the `run`-like utilities is that you get the O(1) conversion to the immutable type without having to use `unsafeFreeze` or similar yourself. I could understand if `STArray.run` were implemented using FFI, but it isn't, it's just a thin wrapper around `ST.run`. It seems a bit of a shame to essentially not support any form of abstraction around this inlining. Although I guess that might require a much cleverer optimizer than what we have now?

**Comment by natefaubion** [comment]
It would require a general purpose inliner, and an annotation to force these functions to inline.

**Comment by milesfrain** [comment]
> this is probably what you were seeing in the book

Yep, looks like the same issue. Some code snippets here: https://github.com/purescript/purescript-st/issues/33

**Comment by mhmdanas** [comment]
> ## Expected behavior
> 
> STRefs are never inlined.

@natefaubion perhaps you mean "actual behavior" instead of "expected behavior"?

**Comment by sjpgarcia** [comment]
This seems to occur because of `inlineST`'s implementation that accounts for an extra abstraction that doesn't exist, for example:
https://github.com/purescript/purescript/blob/81dd8ea73ad8eab9cc56be492fe5883bec59b8d0/src/Language/PureScript/CoreImp/Optimizer/MagicDo.hs#L124-L125
This might've been left over from way back when there was still `Eff`...
